server/halomod_app/utils.py

```diff
@@ -0,0 +1,362 @@
+"""Plotting and driving utilities for halomod."""
+import io
+import logging
+import matplotlib.ticker as tick
+
+from halomod import TracerHaloModel
+from halomod.wdm import HaloModelWDM
+from matplotlib.backends.backend_agg import FigureCanvasAgg
+from matplotlib.backends.backend_pdf import FigureCanvasPdf
+from matplotlib.backends.backend_svg import FigureCanvasSVG
+from matplotlib.figure import Figure
+import re
+import codecs
+import pickle
+
+
+logger = logging.getLogger(__name__)
+
+
+def hmf_driver(cls=TracerHaloModel, previous: [None, TracerHaloModel] = None, **kwargs):
+    if previous is None:
+        return cls(**kwargs)
+    elif "wdm_model" in kwargs and not isinstance(previous, HaloModelWDM):
+        return HaloModelWDM(**kwargs)
+    elif "wdm_model" not in kwargs and isinstance(previous, HaloModelWDM):
+        return TracerHaloModel(**kwargs)
+    else:
+        this = previous.clone(**kwargs)
+
+        # TODO: this is a hack, and should be fixed in hmf
+        # we have to set all _params whose model has been changed to {}
+        # so that they don't get carry-over parameters from other models.
+        for k, v in kwargs.items():
+            if k.endswith("model") and v != getattr(this, k).__class__.__name__:
+                this.update(**{k.replace("model", "params"): {}})
+
+        return this
+
+
+def create_canvas(objects, q: str, d: dict, plot_format: str = "png"):
```

Comment:
this function should now be unnecessary

Reply:
Ended up using it again in #69 out of necessity 😕

Reply:
Hmm, interesting. Does that mean that matplotlib is being used to generate the figure just for a pdf?

Reply:
@steven-murray At the moment, yes that is correct.

Reply:
OK. We can talk about this more this evening, but I wonder if we couldn't just do a hacky "convert" of the `.svg` on the client to `.pdf` and serve it up.